From f5b677895958c51d0bda397dfc533f4a6ee99fa2 Mon Sep 17 00:00:00 2001
From: Stefan Chulski <stefanc@marvell.com>
Date: Wed, 19 Jul 2017 14:24:48 +0300
Subject: [PATCH 1090/1345] fix: net: mvpp2x: Clear aggregated TXQ during port
 stop procedure

commit  1f70ce48acf3122b3a2aeb205fd6a9898389166b from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

Patch fix crashes caused by ring change procedure during traffic.
Calling ring change procedure during traffic cause race between xmit and
TXQ clean in ring change callback.

Fix:
- Transmit all packets from aggregated TXQ to physical TXQ in port
  stop procedure on each CPU.
- Stop transmit hrtimer on each CPU.

Change-Id: Ibff7f43c45e66aebd60e4bc80475a250a9a179e0
Signed-off-by: Stefan Chulski <stefanc@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/41866
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Yuval Caduri <cyuval@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c |   18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c
index f35c723..318cd4e 100644
--- a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c
+++ b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c
@@ -3692,6 +3692,21 @@ void mv_pp2x_start_dev(struct mv_pp2x_port *port)
 	}
 }
 
+/* Clear aggregated TXQ and kill transmit hrtimer */
+static void mv_pp2x_trans_aggr_txq(void *arg)
+{
+	struct mv_pp2x_port *port = arg;
+	struct mv_pp2x_cp_pcpu *cp_pcpu = this_cpu_ptr(port->priv->pcpu);
+	int cpu = smp_processor_id();
+	struct mv_pp2x_aggr_tx_queue *aggr_txq = &port->priv->aggr_txqs[cpu];
+
+	mv_pp2x_tx_timer_kill(cp_pcpu);
+	aggr_txq->sw_count -= aggr_txq->xmit_bulk;
+	aggr_txq->hw_count += aggr_txq->xmit_bulk;
+	mv_pp2x_aggr_txq_pend_desc_add(port, aggr_txq->xmit_bulk);
+	aggr_txq->xmit_bulk = 0;
+}
+
 /* Set hw internals when stopping port */
 void mv_pp2x_stop_dev(struct mv_pp2x_port *port)
 {
@@ -3706,6 +3721,9 @@ void mv_pp2x_stop_dev(struct mv_pp2x_port *port)
 	/* Disable interrupts on all CPUs */
 	mv_pp2x_port_interrupts_disable(port);
 
+	/* Drain aggregated TXQ on all CPU's */
+	on_each_cpu(mv_pp2x_trans_aggr_txq, port, 1);
+
 	mv_pp2x_port_napi_disable(port);
 
 	netif_carrier_off(port->dev);
-- 
1.7.9.5

